qsirecon/interfaces/gradients.py

```diff
@@ -18,33 +18,214 @@
+    bval_df = _find_shells(original_bvals, max_distance)
+
+    # Check that the correct number of input shells are detected
+    if expected_n_input_shells is not None:
+        n_found_shells = len(bval_df["assigned_shell"].unique())
+        if not expected_n_input_shells == n_found_shells:
+            howmuch, newthresh = (
+                ("too many", "higher")
+                if n_found_shells > expected_n_input_shells
+                else ("too few", "lower")
+            )
+            raise Exception(
+                f"Expected to find {expected_n_input_shells} shells in "
+                f"the input data. Instead we found {n_found_shells}. Having "
+                f"{howmuch} shells detected means you may need to adjust the"
+                f"bval_distance_cutoff parameter to be {newthresh}."
+            )
+
+    # Ensure that at lease 1 b>0 and one b=0 shell are selected
+    select_shell_bs = _parse_shell_selection(requested_gradients, bval_df, max_distance)
+
+    # Make sure the shells are unique (no overlap when accounting for allowed distances)
+    shell_distances = pairwise_distances(select_shell_bs.reshape(-1, 1))
+    if np.any(shell_distances[np.triu_indices_from(shell_distances, k=1)] < max_distance):
+        raise Exception(
+            "Shells and bval_distance_cutoff have overlap. Choose a lower "
+            "bval_distance_cutoff or more separated shells."
+        )
+
+    selected_indices = np.flatnonzero(bval_df["assigned_shell"].isin(select_shell_bs))
+
+    return selected_indices
+
+
+def _parse_shell_selection(requested_bvals, bval_df, max_distance):
+    """Turn bval requests into numbers. Add a 0 if none were originally included."""
+    numeric_bvals = []
+    for requested_bval in requested_bvals:
+        if requested_bval == "highest":
+            highest_shell = bval_df["assigned_shell"].max()
+            LOGGER.info(f"Selecting b={highest_shell} as the highest shell")
+            numeric_bvals.append(highest_shell)
+        elif requested_bval == "lowest":
+            lowest_shell = bval_df["assigned_shell"][
+                bval_df["assigned_shell"] > max_distance
+            ].min()
+            LOGGER.info(f"Selecting b={lowest_shell} as the lowest b>0 shell")
+            numeric_bvals.append(lowest_shell)
+        else:
+            # Find the closest detected shell
+            ok_shells = bval_df["assigned_shell"][
+                np.logical_and(
+                    bval_df["assigned_shell"] >= requested_bval - max_distance,
+                    bval_df["assigned_shell"] <= requested_bval + max_distance,
+                )
+            ]
+            if len(ok_shells.unique()) > 1:
+                raise Exception(
+                    f"Unable to unambiguously select b={requested_bval}."
+                    f"instead select from {bval_df.assigned_shells.unique().tolist()}"
+                )
+
+            numeric_bvals.append(ok_shells.iloc[0])
```

Comment:
Shouldn't this check apply across requested bval options?

Reply:
you mean including "highest", "lowest" and "0"? Those should be taken care of by the clustering

Reply:
No I mean if you put in "highest" for a spectrum-type set of bvals, shouldn't it raise an error? But this check (which is the one that raises an error) only applies if you use a number.

Reply:
`_find_shells` has to be called first to create the `bval_df` needed by this function. Since an exception will get raised in `_find_shells` if the scheme isn't shelled, qsirecon won't make it to this function

Reply:
Ohh okay. Thanks for explaining!